### ISSUE TITLE
Rename TryRead::read to TryRead::try_read and TryWrite::write to TryWrite::try_write

### ISSUE THREAD

**Issue by jnicholls** [carllerche/mio#169]
Many streaming APIs have and will continue to standardize on `Read + Write` types.  For example, `SslStream<S>` in the [rust-openssl](https://github.com/sfackler/rust-openssl/blob/master/openssl/src/ssl/mod.rs#L775) crate.  Starting with v0.4.0, Mio will provide its own Io types (`TcpStream`, `UdpSocket`, etc.) that are non-blocking by default, rather than utilizing existing std::net types that are blocking by default.

This change provoked the creation of the `TryRead + TryWrite` traits, which have a `read` and `write` method respectively.  However it's been discussed recently that there is significant value to make the Mio types compatible with existing and future streaming APIs that are `Read + Write`, whether or not they are prepared to handle `ErrorKind::WouldBlock` (`EWOULDBLOCK` and/or `EAGAIN`).  Having the Mio types provide `Read + Write` out of the box would save the pain of Mio users from having to solve this problem on their own via wrapper types that implement `Read + Write` on top of `TryRead + TryWrite`.

One such solution is to [implement `Read + Write` underneath the `TryRead + TryWrite` traits](https://github.com/carllerche/mio/pull/167).  This is clean and simple, but with the current `TryRead::read()` and `TryWrite::write()` trait methods, there is the issue of having to explicitly resolve the conflict with `Read::read()` and `Write::write()` respectively.  Hence, the idea to rename these methods to `TryRead::try_read()` and `TryWrite::try_write()` as a part of 0.4.0's API incompatible updates over 0.3.x.

Another option is to implement `Read + Write` on top of all `TryRead + TryWrite` types, rather than underneath of them; basically, reverting the handling of ErrorKind::WouldBlock and pass that Err() back to the caller.

In either of the above options for providing `Read + Write`, eliminating the method name ambiguity of ::read() and ::write() would facilitate a much friendlier use of both sets of traits in the same module.

**Comment by wycats** [comment]
@jnicholls that seems reasonable, although it seems somewhat annoying that names are "global" in this way (due to trait resolution). Somewhat annoying, but for the most part life.

I give this a :+1:

**Comment by jnicholls** [comment]
Yes, in general this is a shame and might be something that can be solved w/ Trait Specializations; a post-1.0 improvement that Rust developers would like to solve.  I think in the majority of cases, name collisions are okay because they are rare enough to deal with them by explicitly calling the trait methods via `Trait::method()`.  However in the case of `Read + Write` and `TryRead + TryWrite`, they're likely to be paired with one another in the same module on a regularly basis, and thus the name change to avoid the inelegance of explicit trait method invocation.

Thanks for the thumbs up!

**Comment by carllerche** [comment]
cc @rrichardson @reem

**Comment by reem** [comment]
I had reasonable success just implementing `io::{Read, Write}` for the nonblocking types directly and bypassing `{TryRead, TryWrite}` entirely. It's only ever a semantic error to use them incorrectly, and there is a reasonable amount of existing code which will read/write until an error, a paradigm which works fine with nonblocking readers/writers.

**Comment by rrichardson** [comment]
Echoing what @reem said,  std::io::Error ErrorKind includes WouldBlock, so having a result of simply Result<usize> seems perfectly acceptable.  I am for removing the Try* traits entirely.  Any socket created by Mio routines would be made nonblock by default (with changing to blocking as possible method)

**Comment by rrichardson** [comment]
This seems perfectly reasonable to me. 

```rust 
match sock.write(buf) { 
    Ok(n) => println!("wrote {} bytes", n), 
    Err(ErrorKind::WouldBlock) => eventloop.register_opt(sock, Interest::writable()), 
    Err(e) => panic!(e)
}
```

**Comment by reem** [comment]
Unfortunately you actually have to match on `Err(e) if e.kind() == io::ErrorKind::WouldBlock` which hurts the ergonomics ever so slightly.

**Comment by jnicholls** [comment]
More precisely, `Err(ref e) if e.kind() == io::ErrorKind::WouldBlock`, but it's not really that bad.

I agree with removing TryRead and TryWrite personally.  An error is an error.  If WouldBlock wasn't in std::io then that would be a different story.

**Comment by jnicholls** [comment]
I'm going to submit a new PR that removes TryRead and TryWrite for review.

**Comment by carllerche** [comment]
Requiring matching on WouldBlock means that try! doesnt work correctly anymore. WouldBlock is not really an error case, which is why I mapped it to Option here. 

I'm unsure that I would like to lose that.

**Comment by jnicholls** [comment]
WouldBlock absolutely is an error case! From the standpoint of a user, when that occurs the user has to do something about it...i.e. retry the operation or poll again.  I'd much rather see that as an Err than masquerade as Ok and I think everything went through ok.

Regarding try!, this is much better handled as a new macro such as try_nonblock! which returns your Option.  So much better than TryRead and TryWrite in my opinion.

**Comment by rrichardson** [comment]
Try isn't a great option in functions they use read/write anyways because
of the need to handle event interest and buffer management. That extra
option in there just makes my code even uglier. :)

**Comment by jnicholls** [comment]
Yeah sure, to each their own.  My point was more along the lines of "who cares about try! not taking special cases into account, because it's not meant to."

The openssl crate has its own try_ssl_stream! macro that specially handles the mapping of SSL errors.  I instantly thought about that crate's choice to create a separate macro when Carl mentioned try!.

**Comment by carllerche** [comment]
@rrichardson I've been using `try!(sock.read(...))` plenty and handling the errors at the top level by closing the socket. I'm not sure what you mean by "need to handle event interest" since on a socket error, usually I would just clean up the state & kill the connection and buffer management is handled by a drop handle returning the buffer to the pool and what not.

@jnicholls `WouldBlock` is not an error case w/ non blocking sockets. For example, using epoll in edge triggered mode (without oneshot / re-registering) requires reading / writing to the socket until hitting `WouldBlock`. Even with oneshot & re-registering, `WouldBlock` is going to be handled completely differently than any other error.

So far, I'm not convinced about getting rid of `TryRead` / `TryWrite` :smile_cat:

**Comment by wycats** [comment]
@carllerche I totally agree. 

Having worked with non-blocking IO a bunch, I disagree with treating EAGAIN as an error simply because the system does so.

EAGAIN is a result that you encounter regularly and expectedly when working with these APIs, and forcing error handling code to special case them is just wrong. 

The system ends up using errors for expected results largely because of quirks of idiomatic Unix APIs.

**Comment by rrichardson** [comment]
I do agree that there is overloading of the result space in many POSIX APIs. It is the stance of any languages that use Option/Maybe that bugs can be avoided by forcing disambiguation. 

If it is important for some that there be a semantic distinction between wouldblock/eagain and the rest of the result range, then I am for TryRead::try_read.  It allows those that care to distinguish between the two, and those that don't care will use Read impl'd for T : TryRead.

**Comment by jnicholls** [comment]
As long as TryRead and TryWrite are on top of Read and Write, I don't see any issue of keeping them around; then you can have it both ways.  With that said, I still think a macro is a better choice than a trait.

While I agree w/ the sentiment that EWOULDBLOCK is contextually expected to happen and therefore isn't an error in one's mind, depending on how you define an error, I can still argue the opposite.  I.e. I told the system to read 16 bytes and it didn't do it, thereby not fulfilling the ask resulting in an exceptional case that I have to react differently to.

Ergonomically, matching on Ok(None) doesn't tell me at all what happened, only that it actually failed to perform my read/write operation (key word: failed).  I'd have to resort to documentation to know to know that it's a shorter but unexpressive special case to represent WouldBlock.  Matching on WouldBlock at least tells me what actually happened as I'm reading the code, and it's something that users of non-blocking APIs are used to seeing and handling.

I'm playing devil's advocate here.  In reality, match and/or a macro means that with or without the Try traits, the API is equally easy to use.  There isn't any difference in my mind.  It's six one way and half-a-dozen the other.  Read + Write is the only thing that really matters at the end of the day.

I'll update the PR to include the trait methods renamed to try_* for further feedback. Thanks.

**Comment by carllerche** [comment]
Closed by: f057fb9a11ca4257f8ff4fe8cd2258d3187242c7
